src/networking.c

```diff
@@ -1602,29 +1602,81 @@ int _writeToClient(client *c, ssize_t *nwritten) {
             c->sentlen = 0;
         }
     } else {
-        clientReplyBlock *o = listNodeValue(listFirst(c->reply));
-        size_t objlen = o->used;
+        if (listLength(c->reply) > 1) {
```

Comment:
i think we should fine tune this condition. please see my opinion in the discussion section

Reply:
This is the code before I open this PR, why should we change it?

Reply:
Talking about `if (listLength(c->reply) > 1) {`. I think we should apply this enhancement only if first two buffers are no bigger than NET_MAX_WRITES_PER_EVENT. Please see my explanation below.

Reply:
There is a constraint below, in the `while` loop, which makes sure that we don't gather more than `MAX_IOV_SIZE_PER_EVENT` == 8 `clientReplyBlock`s whose maximum size is 16KB, thus the maximum amount of bytes being sent is in the range of [0, 8*16=128KB], I think it's done? or you want to reduce `MAX_IOV_SIZE_PER_EVENT`, like to 4?

Reply:
in the client list  `c->reply` i don't see any limitation over the size of objects . Where do you this 16KB limitation? 

The big profit of this feature is for aggregating small objects. As for the large objects, we need to be careful and preserve previous behaviour.

Regarding MAX_IOV_SIZE_PER_EVENT, I think the opposite, this value is too small and should be bigger, around 256. there are few commands that send multiple small replies that can be aggressively aggregated.  For example run COMMAND command and hunderds of tiny replies from the server that all can be aggregated into a single write/packet.

Reply:
https://github.com/redis/redis/blob/867816003ec214840c18754be9840f1ddf4192d1/src/networking.c#L319-L353

I get the amount of 16KB from `PROTO_REPLY_CHUNK_BYTES`, it's the minimum size of a node of reply list, there could be some nodes whose size is greater than 16KB, but I still can't find the "small objects" you were talking about, I thought you have to fill a node whose minmum size is 16KB and then move to the next one? please inform me if I'm missing something, thanks~

@moticless

Reply:
https://github.com/redis/redis/pull/9934/commits/47aedb0cc222d1ea5ddddefe0c3cbe18194959cc

Reply:
@panjf2000 I actually don't like the MAX_IOV_SIZE_PER_EVENT limit, because we already have the NET_MAX_WRITES_PER_EVENT limit externally, and doing the same in `_writeToClient` looks too weird.

Reply:
@panjf2000 
If I get the code right, the reply `'len` can be only few bytes but it will be copied to allocated chunk of memory, no smaller than `PROTO_REPLY_CHUNK_BYTES`. But this chunk size is irrelevant to our flow - The written length in the reply buffer is still only `len` (in code: `tail->used = len; `)

@sundb 
I agree with you that having another length limitation at `_writeToClient` looks weird. But this way or another we need to limit this feature to aggregating only small replies and avoid try optimizing big ones. i hope you agree with me about this point.

Reply:
@moticless Indeed, and I'm also speculating whether overly large aggregated packets might affect other connections, leading to an overall performance drop.
I ran a test, there is a very significant drop in performance.

command:
```sh
./src/redis-benchmark -n 1000000 info
```

unstable:
```
  throughput summary: 41951.59 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        1.085     0.160     1.047     1.535     2.015    21.471
```

with this pr:
```
  throughput summary: 40728.22 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        1.118     0.168     1.087     1.503     1.983    11.719
```

Reply:
I also ran a benchmark test but got the exact opposite result：

unstable:
```
Summary:
  throughput summary: 44505.77 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        1.088     0.136     1.223     1.415     1.503     2.919
```

use-writev:

```
Summary:
  throughput summary: 46000.27 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        1.054     0.136     1.199     1.335     1.391     2.959
```

Reply:
@panjf2000 Do you use macOS? M1?
I also test on macOS, same as you.
`writev` on `macOS` will bring more performance gains than on linux.

Reply:
No, I ran the benchmark on my Linux server with 8 cores and 32 GB mem.

Reply:
Ohh, Looks like it's time for me to ditch vm, every time my test results are different from everyone else's.
@moticless Can you also conduct a test?

Reply:
This is the result from my another linux server(2 cores and 4GB mem):

unstable:

```
Summary:
  throughput summary: 29266.29 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        1.563     0.304     1.535     2.199     2.879    19.103
```

use-writev:

```
Summary:
  throughput summary: 29857.88 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        1.530     0.288     1.511     2.079     2.735    11.799
```

However, the benchmark results don't seem to be so stabilized.

Reply:
Hi,
I am running inside container, so don't expect shiny numbers :)

In the following scenario the unstable branch gives little better numbers: 
```
P=PR, L=LATEST (unstable)
> redis-benchmark -n 1000000 info
     avg       min       p50       p95       p99       max
P  2.777     1.248     2.719     3.799     4.655    14.527
P  2.755     0.848     2.703     3.751     4.615     9.519
L  2.698     0.952     2.639     3.735     4.471     7.039
L  2.706     0.720     2.639     3.775     4.511     9.143
```

Since I know that command COMMAND is not optimized and reply with many small writes/packets - I expected to see here dramatic improvement:
```
> ./redis-benchmark -n 10000 COMMAND
     avg       min       p50       p95       p99       max
P  26.045     2.344    24.687    45.023    57.215    72.191
L 290.851    13.680   302.847   398.079   419.327   449.279
```

In case of big packets, the numbers are more or less the same:
```
> ./redis-benchmark -n 10000 -d 1000000 -t set,get
     avg       min       p50       p95       p99       max
P   2.242     0.328     2.167     2.543     3.215    29.391
P   2.205     0.872     2.151     2.647     3.247    20.047
L   2.223     0.136     2.191     2.567     3.223    21.183
L   2.270     0.840     2.183     2.599     3.143    36.959
```
Please note that performance will be even worse against the PR in case of TLS, as we are making there additional copy. We need to fine tune the scenario of small packets while all the rest should preserve original flow.

Reply:
... Following review our reply infrastructure, I have an idea that might simplify the solution
- Can be common to TLS and non-TLS 
- and avoid from using writev() (and mocking it for TLS). 

It goes as follows:
- (As we want to optimize only for packets) Each allocated clientReplyBlock (CRB) is usually of `size` much bigger than actually being `used`. 
- if there is more than 1 reply in the list  and there is enough free space in the first object  `(header.size-header.used)`  to include `next.used`, then
```
- totalLen = header.used
- do 
  - memcpy(next.buf, header.buf + totalLen, next.used)
  - totalLen += next.used
  - Advance next
  - until (next == NULL) or (next.used > header.size - totalLen)
- Write (header.buf, totalLen)
- Update c->reply accordingly
```

Reply:
`memcpy(next.buf, header.buf + totalLen, next.used)` should be `memcpy(header.buf + totalLen, next.buf, next.used)`

Reply:
I fail to see how this new approach works better than `writev()`, after all we do more memory copies.

Reply:
any chance that we are able to use `writev()` only when there is a scenario with small packets and as for other cases like big packets and TLS, we just fall back to the approach it used to be?

Reply:
> * (As we want to optimize only for small packets) Each allocated clientReplyBlock (CRB) is usually of `size` much bigger than actually being `used`.
> * if there is more than 1 reply in the list  and there is enough free space in the first object  `(header.size-header.used)`  to include `next.used`, then

    @moticless I don't fully understand this, if the length of replies is greater than 1,
    then the first clientReplyBlock will definitely be filled up to clientReplyBlock->size,
    it won't have any free space.

Reply:
> Please note that performance will be even worse against the PR in case of TLS, as we are making there additional copy. We need to fine tune the scenario of small packets while all the rest should preserve original flow.

I ran a benchmark on Linux and the result seemed to have a different opinion:

unstate:

```
RUN #1] Preparing benchmark client...
[RUN #1] Launching threads now...
[RUN #1 100%,  40 secs]  0 threads:     1000000 ops,   25219 (avg:   24549) ops/sec, 1.05MB/sec (avg: 1.02MB/sec),  0.04 (avg:  0.04) msec latencycycy

1         Threads
1         Connections per thread
1000000   Requests per client


ALL STATS
============================================================================================================================
Type         Ops/sec     Hits/sec   Misses/sec    Avg. Latency     p50 Latency     p99 Latency    p100 Latency       KB/sec
----------------------------------------------------------------------------------------------------------------------------
Sets         2231.78          ---          ---         0.04679         0.04700         0.07100         0.08700       171.93
Gets        22317.52       102.76     22214.75         0.03881         0.03900         0.06300         0.07900       872.77
Waits           0.00          ---          ---             ---             ---             ---             ---          ---
Totals      24549.29       102.76     22214.75         0.03953         0.03900         0.06300         0.07900      1044.70
```

use-writev:

```
[RUN #1] Preparing benchmark client...
[RUN #1] Launching threads now...
[RUN #1 100%,  40 secs]  0 threads:     1000000 ops,   25346 (avg:   24892) ops/sec, 1.06MB/sec (avg: 1.04MB/sec),  0.04 (avg:  0.04) msec latencycy

1         Threads
1         Connections per thread
1000000   Requests per client


ALL STATS
============================================================================================================================
Type         Ops/sec     Hits/sec   Misses/sec    Avg. Latency     p50 Latency     p99 Latency    p100 Latency       KB/sec
----------------------------------------------------------------------------------------------------------------------------
Sets         2263.02          ---          ---         0.04630         0.04700         0.07100         0.08700       174.34
Gets        22629.91       205.34     22424.57         0.03838         0.03900         0.05500         0.07100       888.34
Waits           0.00          ---          ---             ---             ---             ---             ---          ---
Totals      24892.93       205.34     22424.57         0.03910         0.03900         0.06300         0.07900      1062.68
```

Actually, I now tend to think that the fluctuation of bench results between unstable and use-writev is intermittent other than `COMMAND`.

Reply:
@panjf2000 I use `memtier` to test `COMMAND`,  and the results were the opposite of @moticless, which is weird.
unstable
```
===================================================================================================
Type          Ops/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec 
---------------------------------------------------------------------------------------------------
Commands       128.82      1046.52800      1048.57500      1048.57500      1048.57500     29112.66 
Totals         128.82      1046.52800      1048.57500      1048.57500      1048.57500     29112.66
```

use-writev
```
===================================================================================================
Type          Ops/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec 
---------------------------------------------------------------------------------------------------
Commands       504.70       397.42897       397.31100       528.38300       544.76700    114060.39 
Totals         504.70       397.42897       397.31100       528.38300       544.76700    114060.39
```

Reply:
@sundb 
I think you misread his data, actually, your result is consistent with his, `use-writev` is way faster than `unstable` with `COMMAND`, the main disagreement here is a concern that `writev()` may have a negative impact on the performance of other redis commands, but according to my experiences, running benchmark test between `unstable` and `use-writev` didn't always get the stable results: sometimes `use-writev` is better, sometimes `unstable` is better..., therefore I tend to think that's due to other factors and `writev()` is ok to go.

Reply:
@panjf2000 @moticless I ran a test with `COMMAND` and `GET` and `SET` together, and it seems that even with both big and small packets, the small packet gives a performance improvement, after all, `COMMAND` shows a 3x performance improvement.

unstable

```
============================================================================================================================
Type         Ops/sec     Hits/sec   Misses/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec 
----------------------------------------------------------------------------------------------------------------------------
Sets           38.38          ---          ---       477.96215       524.28700       770.04700       937.98300         2.95 
Gets          383.38         0.00       383.38       475.89085       522.23900       745.47100       937.98300        14.94 
Waits           0.00          ---          ---             ---             ---             ---             ---          --- 
Totals        421.76         0.00       383.38       476.07737       522.23900       745.47100       937.98300        17.90 
```

pr
```
============================================================================================================================
Type         Ops/sec     Hits/sec   Misses/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec 
----------------------------------------------------------------------------------------------------------------------------
Sets          101.35          ---          ---       176.92865       161.79100       378.87900       436.22300         7.80 
Gets         1012.35         0.00      1012.35       179.69712       159.74300       387.07100       794.62300        39.46 
Waits           0.00          ---          ---             ---             ---             ---             ---          --- 
Totals       1113.70         0.00      1012.35       179.44513       159.74300       387.07100       794.62300        47.25
```

Reply:
Did you run this benchmark test on Linux or macOS？With TLS or without TLS？

Reply:
@panjf2000 I have only tested ubuntu without TLS.

Reply:
I think we need a more comprehensive test.

* Different packet sizes
  1. small packet
  2. large packet
  3. small packet and large packet mix

* Enable or disable io-threads
* With or without TLS.

Reply:
Yes, there is still a lot to do back there, I will try to accomplish more benchmarks during the weekend, thank you guys for all the efforts here, I really appreciate it.

Reply:
@panjf2000 , following reviewing the code, we can further optimize the code by involve the most often used buffer in writev as well, which is client's static buffer. (i.e. , the client has a static buffer that being reused, before attempting to malloc new buffers)

On note about performance, since time measurements might be rather Inconclusive when it reaches big buffers or optimized commands, alternative approach can simply to count system-calls with `strace`.  Attaching PID and trace write() sys-call: `strace -e trace=write -p 1234`.  Or get statistics per syscall type ` strace -c -p 1234`.

Reply:
> following reviewing the code, we can further optimize the code by involve the most often used buffer in writev as well, which is client's static buffer. (i.e. , the client has a static buffer that being reused, before attempting to malloc new buffers)

You mean add `client->buf` into iovec for `writev()`?

Reply:
Yes. 

I still think that this feature is effective only for small buffers alone. In case of big buffers there are much more considerations that we need to take into account.

Reply:
Please check out the latest commit which includes static buffer in the buffer queue of `writev()`, and we may want to do some new benchmark tests.
@sundb @moticless

Reply:
@oranagra I think you might want to take a look here.

Reply:
i'm not certain what to look at here.
please try to sum up the pending things.
i see two things in that discussion.
1. discussion around the maximum size that can be written using writev.
2. benchmark results, which may no longer be relevant, since i understand the static buffer wasn't part of writev when these benchmarks were run.

regarding the maximum size, is the discussion around fairness or performance?

from fairness point of view, i think the old code wasn't really fair anyway, it attempted to avoid writing more buffers into the socket if it passed NET_MAX_WRITES_PER_EVENT, but if a single buffer contained 500mb, it didn't avoid trying to write as much as the socket could take.

regarding performance, i'm not sure what there is to lose here, i didn't look into the kernels' implementation of writev, but if i assume it is as efficient, and avoids copying data, then all we need to do is:
1. avoid copying data on our side, i don't see any copy on our side apart from the code in TLS, which i think could be improved to only do that for a successive list of very short buffers.
2. if there are are series of large buffers, maybe we can avoid using writev, but i'm not sure it matters.
3. for the common case of only static buffer, after my review, we keep using the old code, that's a very common case.

please let me know which parts of this long discussion are still pending and what would you like me to respond to.